### ISSUE TITLE
Proposal for a Screenshot-Capturing API in Browsers

### ISSUE THREAD

**Issue by eladalon1983** [niklasvh/html2canvas#2509]
Hello all,

I'm one of the developers working for Chrome. I'm drafting a proposal for introducing an API to allow grabbing a screenshot. (This would probably cover some of the use cases `html2canvas` solves, but not all, as it will require manual user confirmation, which IIUC `html2canvas` does not require.)

How can you help?
1. You could check out my [very early draft](https://eladalon1983.github.io/mediacapture-screenshot/) and leave some [comments](https://github.com/eladalon1983/mediacapture-screenshot/issues) and/or PRs.
2. When the time comes to ship this, if consensus cannot be reached with other browser vendors (a decently likely outcome), the other option is to show overwhelming interest from web developers. If you register interest by leaving a comment either here or on my [repo](https://github.com/eladalon1983/mediacapture-screenshot/issues), that would be a good first step.

Thanks,
Elad

**Comment by bripkens** [comment]
I just want to say thanks for picking this up @eladalon1983! I repeatedly came across this use case and the lack of a standardized solutions has always hindered productization/has stopped initiatives.

**Comment by bripkens** [comment]
Leveraging the tagged StackOverflow questions might also serve to validate interest: https://stackoverflow.com/questions/tagged/html2canvas?sort=newest

**Comment by TomasHubelbauer** [comment]
Hey @eladalon1983 and I think also @niklasvh: I have developed an experimental alternative to `html2canvas` which uses the browser screen share API (`getDisplayMedia`) and uses the screen share stream to capture a screenshot and crop out a desired area out of it. You can find it here: https://github.com/TomasHubelbauer/selfie

It is not a completely flawless solution, from my side it was an experiment I wanted to see the full potential of. What the flow boils down to is this:

- Let the user know you're going to ask them a permission for the screenshare to capture the self-screenshot
- Direct them to either select own tab or a full screen screenshare (so the pages surface is in the shared stream)
- Use any way to crop out the desired element's area from the screenshot (my library's approach is one of many)

I think this flow asks too much of the user, so I'm not completely happy with this, but folks might be interested in this as a fallback to `html2canvas` for cases where it doesn't work well for them (I've had some troubles with SVGs) and it might also help support your argument @eladalon1983 for an inclusion of a self-screenshot API in the browsers.

I'd be happy to chat about this if anyone is interested in any aspect of the library or the approach it uses.

**Comment by eladalon1983** [comment]
Apologies for being late with an update. At the moment it looks like a Screenshot API would only be possible if subject to (1) cross-origin isolation and (2) a yet-to-be-specified opt-in header. These are not widely adopted yet (especially number 2), so work on a screenshot API is taking a back seat to some other projects.

### **But I have good news!**

I have recently introduced a different Chrome feature, currently in [origin trial](https://github.com/GoogleChrome/OriginTrials/blob/gh-pages/developer-guide.md), and it gets you much closer to what you need. It's called [getCurrentBrowsingContextMedia](https://groups.google.com/a/chromium.org/g/blink-dev/c/j70Xe2xSh1s/m/XeQ5UIr8AgAJ), but no worries, the name is not going to stick.

Please see a [**demo** here](https://eladalon1983.github.io/gcbcm-demo/). Works on Chrome m90-m92.

TL;DR is that it brings up a browser media-picker that looks like this:
![Screen Shot 2021-06-17 at 18 48 40](https://user-images.githubusercontent.com/22117736/122440485-e6440f80-cf9c-11eb-9c28-c0c14564783a.png)

Demo-caveats:
* For posterity - this demo will work for a few weeks, after which the OT token will expire.
* This demo was hastily thrown together by a C++ developer. It ain't pretty. Feel free to submit a PR to improve it.

If you find this API useful, please say so publicly on [this thread](https://github.com/WICG/proposals/issues/32), preferably along with the information of which company you work for. We can only ship this if we get **positive** developer feedback, so if this is useful for you, you **need** to leave a message there. :-)

**Comment by eladalon1983** [comment]
Sorry for being long winded. Here's the TL;DR:
[Demo here](https://eladalon1983.github.io/gcbcm-demo/), Chrome m90-m92.
Will only be adopted if positive feedback is posted [here](https://github.com/WICG/proposals/issues/32).

**Comment by kuige** [comment]
In some situations, I think node capture API is more helpful, for example, `document.body.captureAsImage()` will return the image capture of body

**Comment by eladalon1983** [comment]
> In some situations, I think node capture API is more helpful, for example, `document.body.captureAsImage()` will return the image capture of body

Correct me if I am wrong, but this is a hypothetical API, and not an existing one? If so, you're welcome to suggest it, implement it and drive standardization of it, but I am afraid you'll find it easier said than done.  But [preferCurrentTab] is already implemented and Chrome is a third of the way into approving shipping it. Better one bird in the hand than two on the tree, they say where I am from. :-)
